app/gui2/src/util/config.ts

```diff
@@ -0,0 +1,477 @@
+export interface AnyGroup {
+  name: string
+  description: string
+  options: OptionsRecord
+  groups: GroupsRecord
+  setPath(path: string[]): void
+  clone(): this
+  merge<Other extends AnyGroup>(other: Other): this & Other
+  load(config: StringConfig, stack?: string[]): string[]
+  stringify(): StringConfig
+  prettyPrint(indent: number): string
+  optionsRecursive(): AnyOption[]
+}
+
+/** Options group. Allows defining nested options. The class is generic in order to allow TypeScript
+ * to infer the types of its children and thus allow accessing them in a type-safe way. */
+export class Group<Options extends OptionsRecord, Groups extends GroupsRecord> {
+  name = 'unnamed'
+  description: string
+  options: Options = {} as Options
+  groups: Groups = {} as Groups
+  constructor(cfg?: {
+    description?: string | undefined
+    options?: Options | undefined
+    groups?: Groups | undefined
+  }) {
+    this.description = cfg?.description ?? 'No description.'
+    const options = cfg?.options
+    if (options != null) {
+      for (const [name, option] of Object.entries(options)) {
+        this.addOption(name, option)
+      }
+    }
+    const groups = cfg?.groups
+    if (groups != null) {
+      for (const [name, group] of Object.entries(groups)) {
+        this.addGroup(name, group)
+      }
+    }
+  }
+
+  addOption(name: string, option: AnyOption) {
+    const existingOption = this.options[name]
+    if (existingOption != null) {
+      console.error(`Duplicate config option found '${existingOption.qualifiedName()}'.`)
+    }
+    const options = this.options as OptionsRecord
+    options[name] = option
+    option.name = name
+  }
+
+  addGroup(name: string, group: AnyGroup) {
+    const existingGroup = this.groups[name]
+    if (existingGroup != null) {
+      existingGroup.merge(group)
+    } else {
+      const groups = this.groups as GroupsRecord
+      groups[name] = group
+      group.name = name
+    }
+    group.setPath([name])
+  }
+
+  /** Set the path of this group. If this group was placed in another group, the path will contain
+   * all the parent group names. */
+  setPath(path: string[]) {
+    for (const option of Object.values(this.options)) {
+      option.path = path
+    }
+    for (const [name, group] of Object.entries(this.groups)) {
+      group.setPath(path.concat([name]))
+    }
+  }
+
+  /** Return a deep copy of this group definition. */
+  clone(): this {
+    const result: AnyGroup = new Group()
+    for (const [name, group] of Object.entries(this.groups)) {
+      result.groups[name] = group.clone()
+    }
+    for (const [name, option] of Object.entries(this.options)) {
+      result.options[name] = option.clone()
+    }
+    return result as this
+  }
+
+  /** Merge this group definition with another group definition. Returns a deeply merged group. In
+   * case the argument will override some options, errors will be logged. */
+  merge<Other extends AnyGroup>(other?: Other | null): this & Other {
```

Comment:
Is this ever used?

Reply:
i think only when parsing CLI options, so should be safe to remove here.